axlearn/open_api/metrics/tool_use_execution.py

```diff
@@ -232,6 +249,128 @@ def _compare_tool_calls(
     return True
 
 
+@dataclasses.dataclass
+class DetailedMatchResult:
+    """Represents the tool matches for different metrics."""
+
+    func_name_match: bool = False
+    strict_arg_match: bool = False
+    lenient_arg_match: bool = False
+    lenient_bow_arg_match: bool = False
+
+
+def _compare_tool_call_detailed(
+    *,
+    pred_tool_calls: Sequence[dict[str, Any]],
+    target_tool_calls: Sequence[dict[str, Any]],
+) -> list[DetailedMatchResult]:
+    """Performs a detailed comparison of the predicted tool calls with the target tool calls and
+    returns different metrics.
+
+    The number of predicted and target tool calls can be different. The comparsion works by
+    iterating over all predicted calls and finding the corresponding target call. Correspondence
+    is defined here by the following rules:
+    - Strict metric: The function name matches and all arguments are identical.
+    - Lenient metric: The function name matches and all arguments match by the lenient argument
+      comparision.
+    - Lenient BOW metric: The function name matches and all arguments match by the lenient BOW
+      comparison.
+
+    Note, that this function return the matching results for every predicted tool call. The final
+    metric is taking the overall number of target calls into account too. That happens in the
+    calling function.
+
+    Args:
+        pred_tool_calls: Predicted tool calls.
+        target_tool_calls: Target tool calls.
+
+    Returns:
+        DetailedMatchResult, containing the Booleans denoting matches for different metrics.
+    """
+
+    target_tool_calls = list(target_tool_calls)
+    pred_tool_calls = list(pred_tool_calls)
+
+    target_funcs = []
+    for t in target_tool_calls:
+        if not "function" in t:
+            continue
+        try:
+            t["function"]["arguments"] = _extract_arguments(t["function"])
+        except (json.JSONDecodeError, KeyError):
+            logging.error("Unable to decode arguments from target call %s", t["function"])
+            continue
+        target_funcs.append(t["function"])
+
+    strict_target_funcs = list(target_funcs)
+    lenient_target_funcs = list(target_funcs)
+    lenient_bow_target_funcs = list(target_funcs)
+
+    results = []
+    for pred_tool in pred_tool_calls:
+        pred_func = pred_tool["function"]
+        try:
+            pred_func["arguments"] = _extract_arguments(pred_func)
+        except (json.JSONDecodeError, KeyError):
+            logging.error("Unable to decode arguments from predicted call %s", pred_func)
+            continue
+
+        fname_match = False
+        strict_match = False
+        lenient_match = False
+        lenient_bow_match = False
+
+        for idx, target_func in enumerate(strict_target_funcs):
+            if pred_func["name"] == target_func["name"]:
+                fname_match = True
+            else:
+                continue
```

Comment:
The definition of `DetailedMatchResult` suggests that a function can match on args but not function name, for example:
```
DetailedMatchResult(
    func_name_match=False,
    strict_arg_match=True,
    ...
)
```
If this configuration can never be the case, we should clarify the behavior somewhere.

Reply:
Correct. Clarified. Its actually an internal and is not exposed to the user.

Reply:
Thanks for clarifying. Please see other comment re. making `DetailedMatchResult` private.